conngen/conngen.cpp

```diff
@@ -37,185 +34,312 @@
-  if ( targets == NULL )
+  int source, target, num_parameters = cg->arity();
+  if ( num_parameters == 0 )
   {
-    LOG( M_ERROR, "CGConnect_cg_i_i_D_l", "targets must be a subnet." );
-    throw SubnetExpected();
+    // connect source to target
+    while ( cg->next( source, target, NULL ) )
+    {
+      // No need to check for locality of the target node, as the mask
+      // created by cg_set_masks() only contain local nodes.
+      Node* const target_node =
+        kernel().node_manager.get_node( target_gids[ target ] );
+      const thread target_thread = target_node->get_thread();
+      kernel().connection_manager.connect(
+        source_gids[ source ], target_node, target_thread, synmodel_id );
+    }
   }
-  if ( not targets->is_homogeneous() )
+  else if ( num_parameters == 2 )
   {
-    LOG( M_ERROR,
-      "CGConnect_cg_i_i_D_l",
-      "targets must be a homogeneous subnet." );
-    throw BadProperty();
+    if ( not params_map->known( names::weight )
+      or not params_map->known( names::delay ) )
+    {
+      throw BadProperty(
+        "The parameter map has to contain the indices of weight and delay." );
+    }
+
+    long w_idx = ( *params_map )[ names::weight ];
+    long d_idx = ( *params_map )[ names::delay ];
+    std::vector< double > params( 2 );
+
+    // connect source to target with weight and delay
+    while ( cg->next( source, target, &params[ 0 ] ) )
+    {
+      // No need to check for locality of the target node, as the mask
+      // created by cg_set_masks() only contain local nodes.
+      Node* const target_node =
+        kernel().node_manager.get_node( target_gids[ target ] );
+      const thread target_thread = target_node->get_thread();
+      kernel().connection_manager.connect( source_gids[ source ],
+        target_node,
+        target_thread,
+        synmodel_id,
+        params[ d_idx ],
+        params[ w_idx ] );
+    }
   }
-  if ( dynamic_cast< Subnet* >( *targets->local_begin() ) )
+  else
   {
     LOG( M_ERROR,
-      "CGConnect_cg_i_i_D_l",
-      "Only 1-dim subnets are supported as targets." );
-    throw BadProperty();
+      "CGConnect",
+      "Either two or no parameters in the ConnectionSet expected." );
+    throw DimensionMismatch();
   }
+}
 
-  const Token synmodel =
-    kernel().model_manager.get_synapsedict()->lookup( synmodel_name );
-  if ( synmodel.empty() )
-  {
-    throw UnknownSynapseType( synmodel_name.toString() );
-  }
-  const index synmodel_id = static_cast< index >( synmodel );
+/**
+ * Set the masks on the ConnectionGenerator cg. This function also
+ * creates the masks from the given RangeSets sources and targets.
+ *
+ * \param cg The ConnectionGenerator to set the masks on
+ * \param sources The source ranges to create the source masks from
+ * \param targets The target ranges to create the target masks from
+ */
+void
+cg_set_masks( ConnectionGeneratorDatum& cg,
+  RangeSet& sources,
+  RangeSet& targets )
```

Comment:
As far as I can see, `sources` and `targets` are not modified below. Why not pass as `const&`?

Reply:
They are `const GIDCollection&` now. See fe4408921.